src/data/index.ts

```diff
@@ -89,22 +166,91 @@ export class Index<T extends RecordMetadata = RecordMetadata> {
     this._upsertCommand = new UpsertCommand<T>(apiProvider, namespace);
   }
 
+  /**
+   * Returns an {@link Index} targeting the specified namespace. By default, all operations take place inside the default namespace ''.
+   *
+   * @example
+   * ```js
+   * const pinecone = new Pinecone();
+   * const indexNamed = pinecone.index('my-index').namespace('my-namespace');
+   * ```
+   *
+   * @param namespace - The namespace to target within the index.
+   */
   namespace(namespace: string): Index<T> {
     return new Index<T>(this.target.index, this.config, namespace);
   }
 
+  /**
+   * Upsert records to the index.
+   *
+   * @example
+   * ```js
+   * const pinecone = new Pinecone();
+   * await pinecone.index('my-index').upsert([{
+   *  id: 'record-1',
+   *  values: [0.176, 0.345, 0.263],
+   * },{
+   *  id: 'record-2',
+   *  values: [0.176, 0.345, 0.263],
+   * }])
+   * ```
+   *
+   * @param data - An array of {@link PineconeRecord} objects to upsert.
+   * @returns A promise that resolves when the upsert is completed.
+   * @throws {@link Errors.PineconeConnectionError} when invalid environment, project id, or index name is configured.
+   * @throws {@link Errors.PineconeArgumentError} when invalid arguments are passed.
```

Comment:
I'm unsure whether we want to be annotating the different `@throws`, but I went ahead and added them for now.

Reply:
Seems useful. I wish typescript supported `throws` natively.